src/uct/tcp/tcp_listener.c

```diff
@@ -0,0 +1,216 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "tcp_listener.h"
+#include <ucs/sys/sock.h>
+#include <ucs/async/async.h>
+
+
+static void uct_tcp_listener_client_data_handler(int fd, void *arg)
+{
+}
+
+static void uct_tcp_listener_conn_req_handler(int fd, void *arg)
+{
+    uct_tcp_listener_t *listener = (uct_tcp_listener_t *)arg;
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    struct sockaddr_storage client_addr;
+    ucs_async_context_t *async_ctx;
+    uct_tcp_listener_ctx_t *listen_ctx;
+    ucs_status_t status;
+    socklen_t addrlen;
+    int accept_fd;
+
+    ucs_assert(fd == listener->listener_fd);
+
+    for (;;) {
+        addrlen = sizeof(struct sockaddr_storage);
+
+        accept_fd = accept(listener->listener_fd, (struct sockaddr*)&client_addr,
+                           &addrlen);
+        if (accept_fd == -1) {
+            if ((errno != EAGAIN) && (errno != EWOULDBLOCK) && (errno != EINTR)) {
+                ucs_error("accept() failed: %m");
+            }
+            return;
+        }
+
+        ucs_trace("server accepted a connection request from client %s",
+                  ucs_sockaddr_str((struct sockaddr*)&client_addr, ip_port_str,
+                                   UCS_SOCKADDR_STRING_LEN));
+
+        /* Set the accept_fd to non-blocking mode
+         * (so that send/recv won't be blocking) */
+        status = ucs_sys_fcntl_modfl(accept_fd, O_NONBLOCK, 0);
+        if (status != UCS_OK) {
+            close(accept_fd);
+            return;
+
+        }
+
+        listen_ctx = ucs_malloc(sizeof(uct_tcp_listener_ctx_t), "accept listen_ctx");
+        if (listen_ctx == NULL) {
+            ucs_error("failed to allocate memory for a listen_ctx");
+            close(accept_fd);
+            return;
+        }
+
+        listen_ctx->fd  = accept_fd;
+        async_ctx       = listener->super.cm->iface.worker->async;
+
+        UCS_ASYNC_BLOCK(async_ctx);
+        ucs_list_add_tail(&listener->fd_list, &listen_ctx->list);
```

Comment:
why this list is needed? why don't invoke user callback in async mode here instead of `ucs_async_set_event_handler(uct_tcp_listener_client_data_handler)`?

Reply:
the list is used to remove all the used accept_fds in the handler and to free the listen_ctx in the listener cleanup. 
the handler will be used to handle all the incoming data from the client - both data it sends during wireup and events on the ep such as disconnect.
are you suggesting not to use an async handler here?

Reply:
IMO, if you are not going to implement SYNC callback mode, I don't see the reason use async context here